lib/buffer.js

```diff
@@ -446,7 +446,48 @@ Buffer.prototype.compare = function compare(b) {
   return binding.compare(this, b);
 };
 
-function slowIndexOf(buffer, val, byteOffset, encoding) {
+
+// Finds either the first index of `val` in `buffer` at offset >= `byteOffset`,
+// OR the last index of `val` in `buffer` at offset <= `byteOffset`
+//
+// Arguments:
+// - buffer - a Buffer to search
+// - val - a string, Buffer, or number
+// - byteOffset - an index into `buffer`; will be clamped to an int32
+// - encoding - an optional encoding, relevant is val is a string
+// - dir - true for indexOf, false for lastIndexOf
+function bidirectionalIndexOf(buffer, val, byteOffset, encoding, dir) {
+  if (typeof byteOffset === 'string') {
+    encoding = byteOffset;
+    byteOffset = undefined;
+  } else if (byteOffset > 0x7fffffff) {
+    byteOffset = 0x7fffffff;
+  } else if (byteOffset < -0x80000000) {
+    byteOffset = -0x80000000;
+  }
+  if (typeof byteOffset !== 'number' || isNaN(byteOffset)) {
```

Comment:
I think we need this to match the behavior of `Buffer.indexOf`

* `buf.indexOf('foo')` searches the whole buffer, as does `buf.indexOf('foo', null)`, `buf.indexOf('foo', 'foo')`, etc
* `buf.indexOf('foo', 0)` searches starting from index 0, which also searches the whole buffer
* `buf.lastIndexOf('foo')` should def search the whole buffer, but
* `buf.lastIndexOf('foo', 0)` does a *reverse* search starting from index 0, so it only checks for a match at index 0

So a minimum, we have to special-case `undefined`

I think it's best if `buf.lastIndexOf('foo', null)`, `buf.lastIndexOf('foo', NaN)` etc match `buf.lastIndexOf('foo')` -- in other words,  they should search the whole buffer. That means they're NOT equivalent to `buf.lastIndexOf('foo', 0)`